src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetRemoteFileService.cs

```diff
@@ -0,0 +1,246 @@
+using System.Net.Http;
+using System.Runtime.Caching;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.ServiceHub.Framework.Services;
+using NuGet.Common;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Protocol;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public sealed class NuGetRemoteFileService : INuGetRemoteFileService
+    {
+        private ServiceActivationOptions? _options;
+        private IServiceBroker _serviceBroker;
+        private AuthorizationServiceClient? _authorizationServiceClient;
+        private bool _disposedValue;
+
+        // TODO: right settings for cache and policy?
+        internal readonly static MemoryCache IdentityToUriCache = new MemoryCache("PackageSearchMetadata",
+            new NameValueCollection
+            {
+                { "cacheMemoryLimitMegabytes", "4" },
+                { "physicalMemoryLimitPercentage", "0" },
+                { "pollingInterval", "00:02:00" }
+            });
+        private static readonly CacheItemPolicy CacheItemPolicy = new CacheItemPolicy
+        {
+            SlidingExpiration = ObjectCache.NoSlidingExpiration,
+            AbsoluteExpiration = ObjectCache.InfiniteAbsoluteExpiration,
+        };
+
+        public static void AddIconToCache(PackageIdentity packageIdentity, Uri uri)
+        {
+            if (uri != null)
+            {
+                IdentityToUriCache.Add("icon:" + packageIdentity.ToString(), uri, CacheItemPolicy);
+            }
+        }
+
+        public static void AddLicenseToCache(PackageIdentity packageIdentity, Uri embeddedLicenseUri)
+        {
+            Assumes.NotNull(embeddedLicenseUri);
+            IdentityToUriCache.Add("license:" + packageIdentity.ToString(), embeddedLicenseUri, CacheItemPolicy);
+        }
+
+        public NuGetRemoteFileService(
+            ServiceActivationOptions options,
+            IServiceBroker serviceBroker,
+            AuthorizationServiceClient authorizationServiceClient)
+        {
+            _options = options;
+            _serviceBroker = serviceBroker;
+            _authorizationServiceClient = authorizationServiceClient;
+
+            Assumes.NotNull(_serviceBroker);
+            Assumes.NotNull(_authorizationServiceClient);
+        }
+
+        public NuGetRemoteFileService(IServiceBroker serviceBroker)
+        {
+            _serviceBroker = serviceBroker;
+            Assumes.NotNull(_serviceBroker);
+        }
+
+        public async ValueTask<Stream?> GetPackageIconAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(packageIdentity);
+            string key = "icon:" + packageIdentity.ToString();
+            Uri? uri = IdentityToUriCache.Get(key) as Uri;
+
+            if (uri == null)
+            {
+                return null;
+            }
```

Comment:
Is it expected that if the cache item is no longer in the cache that getting the icon will no longer work?  Should it fall back to searching the local package sources for the package?

What if the API also had a parameter `string source`? That, coupled with my other comment's suggestion of adding a `GetPackageReaderAsync` method to `FindLocalPackageResource` could eliminate the need for `IdentityToUriCache` and adding `PackagePath` to the public API (although I'm ok with it being added to `LocalPackageInfo`, it totally makes sense from a public API point of view. It doesn't make as much sense for `ClonedPackageSearchMetadata`, but that's a whole other issue).

Reply:
we had considered passing more URL based parameter as part of the contract. 
However, we were worried of the security hole that would open up.

i don't think it will hit null in practice due to a delay of trying to hit the cache.